File: C++/evalutate_reverse_polish_notation.cpp
// no. 150
#include <string>
#include <vector>
using namespace std;
class Solution {
public:
  vector<int> stack;
  vector<int> get_operate_nums() {
    int num1 = stack.back();
    stack.pop_back();
    int num2 = stack.back();
    stack.pop_back();
    vector<int> nums = {num1, num2};
    return nums;
  }
  int evalRPN(vector<string> &tokens) {
    for (string i : tokens) {
      if (i == "+") {
        vector<int> nums = get_operate_nums();
        stack.push_back(nums[0] + nums[1]);
      } else if (i == "-") {
        vector<int> nums = get_operate_nums();
        stack.push_back(nums[1] - nums[0]);
      } else if (i == "*") {
        vector<int> nums = get_operate_nums();
        stack.push_back(nums[0] * nums[1]);
      } else if (i == "/") {
        vector<int> nums = get_operate_nums();
        stack.push_back(nums[1] / nums[0]);
      } else {
        stack.push_back(stoi(i));
      }
    }
    return stack[0];
  }
};

File: python/evalutate_reverse_polish_notation.py
#no. 150

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        def get_operate_nums():
            num1 = stack.pop()
            num2 = stack.pop()
            return num1, num2

        for i in tokens:
            if i == '+':
                num1, num2 = get_operate_nums()
                stack.append(num1 + num2)
            elif i == '-':
                num1, num2 = get_operate_nums()
                stack.append(num2 - num1)
            elif i == '*':
                num1, num2 = get_operate_nums()
                stack.append(num1 * num2)
            elif i == '/':
                num1, num2 = get_operate_nums()
                q = int(num2 / num1)
                stack.append(q)
            else:
                stack.append(int(i))
        return stack[0]
